ada_feeding_perception/ada_feeding_perception/table_detection.py

```diff
@@ -0,0 +1,483 @@
+        except ValueError as err:
+            self.get_logger().warn(
+                f"Error getting type of image topic. Defaulting to CompressedImage. {err}"
+            )
+            image_type = CompressedImage
+        self.image_subscriber = self.create_subscription(
+            image_type,
+            image_topic,
+            self.image_callback,
+            1,
+            callback_group=MutuallyExclusiveCallbackGroup(),
+        )
+
+    def fit_table(
+        self,
+        image_msg: Image,
+        image_depth_msg: Image,
+    ):
+        """
+        Fits a plane to the table based on the given RGB image and depth image messages.
+
+        Parameters
+        ----------
+        image_msg: The RGB image to detect plates from.
+        image_depth_msg: The depth image corresponding to the RGB image.
+
+        Returns
+        ----------
+        table: The depth array of the table.
+        """
+        # Convert ROS images to CV images
+        image = ros_msg_to_cv2_image(image_msg, self.bridge)
+        gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
+
+        image_depth = ros_msg_to_cv2_image(image_depth_msg, self.bridge)
+
+        # Detect Largest Circle (Plate)
+        circles = cv2.HoughCircles(
+            gray,
+            cv2.HOUGH_GRADIENT,
+            self._hough_accum,
+            self._hough_min_dist,
+            param1=self._hough_param1,
+            param2=self._hough_param2,
+            minRadius=self._hough_min,
+            maxRadius=self._hough_max,
+        )
+        if circles is None:
+            return None, None, None
+        circles = np.round(circles[0, :]).astype("int")
+        plate_uv = (0, 0)
+        plate_r = 0
+        print(len(circles))
+        for x, y, r in circles:
+            print("Radius: " + str(r))
+            if r > plate_r:
+                plate_uv = (x, y)
+                plate_r = r
+
+        # Testing - Draw out the circles
+        if circles is not None:
+            circles = np.uint16(np.around(circles))
+            for x, y, r in circles:
+                center = (x, y)
+                # circle center
+                cv2.circle(gray, center, 1, (0, 100, 100), 3)
+                # circle outline
+                radius = r
+                cv2.circle(gray, center, radius, (255, 0, 255), 3)
+                cv2.circle(gray, center, radius + self._table_buffer, (255, 0, 255), 3)
+            # The following lines of code are commented to prevent interruption of the program
+            # while running. Uncomment to view images of the detected circles.
+            # cv2.imshow("detected circles", gray)
+            # cv2.waitKey(0)
+
+        # Create Mask for Depth Image
+        plate_mask = np.zeros(image_depth.shape)
+        cv2.circle(plate_mask, plate_uv, plate_r + self._table_buffer, 1.0, -1)
+        cv2.circle(plate_mask, plate_uv, plate_r, 0.0, -1)
+        depth_mask = (image_depth * (plate_mask).astype("uint16")).astype(float)
```

Comment:
Hmm why are you converting it to type float? Because the depth mask is in integer mm units, right? I'm not sure why the float conversion is necessary.